src/backend/access/transam/xlog.c

```diff
@@ -6763,24 +6779,37 @@ GetLastWrittenLSNv(RelFileLocator relfilenode, ForkNumber forknum,
 			key.blockNum = blkno + i;
 
 			entry = hash_search(lastWrittenLsnCache, &key, HASH_FIND, NULL);
-
 			if (entry != NULL)
+			{
 				lsns[i] = entry->lsn;
+			}
 			else
 			{
-				XLogRecPtr lsn;
-				lsns[i] = lsn = XLogCtl->maxLastWrittenLsn;
+				lsns[i] = InvalidXLogRecPtr;
+				missed_keys = true;
+			}
+		}
+		if (missed_keys)
+		{
+			LWLockRelease(LastWrittenLsnLock);
+			LWLockAcquire(LastWrittenLsnLock, LW_EXCLUSIVE);
 
-				LWLockRelease(LastWrittenLsnLock);
-				SetLastWrittenLSNForBlock(lsn, relfilenode, forknum, key.blockNum);
-				LWLockAcquire(LastWrittenLsnLock, LW_SHARED);
+			lsn = XLogCtl->maxLastWrittenLsn;
+
+			for (int i = 0; i < nblocks; i++)
+			{
+				if (lsns[i] == InvalidXLogRecPtr)
+				{
+					SetLastWrittenLSNForBlockRangeInternal(lsn, relfilenode, forknum, key.blockNum, 1);
+				}
```

Comment:
We need to add the missing block, not just the first one. In passing, we should probably not be using a full function-scoped variable, but just a local one (or alternatively pass the parameter directly - it's locked anyway, so no other backend can update it anyway)

```suggestion
				if (lsns[i] == InvalidXLogRecPtr)
				{
					SetLastWrittenLSNForBlockRangeInternal(XLogCtl->maxLastWrittenLsn, relfilenode, forknum, blkno + i, 1);
				}
```